enhancements/network/ovn-kubernetes-ipsec.md

```diff
@@ -0,0 +1,298 @@
+an approximation of the expected performance degradation - These results will be
+posted at a later date. After implementation of the enhancement, it is expected
+that comprehensive benchmarking and profiling will take place. Based on the
+output of this activity, it may be decided to start some optimization activities
+in parallel to improve performance.
+
+#### Security Review
+
+See "Open Questions" below.
+
+## Design Details
+
+### Open Questions [optional]
+
+- Will this need to go through some kind of Security Review?
+
+### Test Plan
+
+It will expected to add a number of Continuous Integration jobs for IPsec:
+
+- All Openshift e2e tests should be replicated in a cluster that has been
+  installed with IPsec enabled. This should ensure IPsec has no regressions with
+  respect to a cluster that has not been enabled with IPsec.
+- In order to test any issues with version skew, an IPsec-enabled cluster should
+  be upgraded while active.
+
+Additional CI jobs could potentially be added at a later stage.
+
+- Functional tests
+    - Ensure certificate expiration and rotation happens correctly.
+- Non-functional tests
+    - Performance tests (scale tests)
+    - Endurance tests (traffic generation over a long period)
+    - Recovery tests (what happens when node IPsec components such as pluto or
+      ovs-monitor-ipsec are killed)
+
+It should also be noted that, some upstream OVS/OVN unit and integration tests
+will also need to be developed.
+
+### Graduation Criteria
+
+TBD
+
+#### Examples
+
+##### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+- Performance measurement
+
+##### Tech Preview -> GA
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+
+##### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+Currently, it is proposed that this feature can only be enabled at cluster
+installation time.
+
+### Version Skew Strategy
+
+It is proposed that this enhancement will not support mixed clusters of RHEL7
+and RHEL8 nodes. This can be supported in a later enhancement.
+
+However, it will need to support version skew within RHEL8 components. The exact
+details of this will be understood, when it is determined how IPsec components
+will be containerized but some considerations are noted below:
+
+- Internode libreswan compatibility
+  - Each nodes' libreswan version will need to be compatible with other nodes'
+    libreswan version
+- Internode OVN/OVS compatibility
+  - Each nodes' OVN/OVS version will need to be compatible with other nodes'
+    OVS/OVN version. While this will generally be covered by OVN Kubernetes as a
+    whole, there may be incompatibilities between IPsec specific
+    subsystems, for example ovsdb schemas entries of newer OVN/OVS versions.
```

Comment:
Presumably there will not be any incompatible schema changes after the version of OVN/OVS in the first version of OCP that supports IPsec?

Reply:
This has prompted me to think about this a bit more. I don't think OVS has an issue because 1) all the IKE traffic is handled by libreswan and 2) esp data traffic would be unencrypted before it even hits OVS so OVS version skew should not affect IPsec. The one exception, that you reference, is if some nodes didn't support IPsec. However, as this enhancement will only enable IPsec at installation time, this is also not an issue.

There could be an issue if an OVN chassis was incompatible with OVN database nodes (e.g. OVNSB) but I guess this is covered by OVNKubernetes somehow? @danwinship can you confirm this?

Reply:
Yeah, we control all the OVS/OVN software that's deployed. The only time there can be version skew is during a cluster upgrade, and if we were going to depend on some new OVN feature we'd have to be able to handle the possible version skew correctly regardless of IPsec.

Reply:
Removed